cql/src/main/java/org/apache/cassandra/stargate/transport/internal/messages/AuthResponse.java

```diff
@@ -67,25 +67,35 @@ public AuthResponse(byte[] token) {
 
   @Override
   protected CompletableFuture<? extends Response> execute(long queryStartNanoTime) {
-    try {
-      Authenticator.SaslNegotiator negotiator = ((ServerConnection) connection).getSaslNegotiator();
-      byte[] challenge = negotiator.evaluateResponse(token);
-      if (negotiator.isComplete()) {
-        AuthenticatedUser authenticatedUser = negotiator.getAuthenticatedUser();
-        persistenceConnection().login(authenticatedUser);
-        if (authenticatedUser.token() != null) {
-          ((ServerConnection) connection).clientInfo().setAuthToken(authenticatedUser.token());
-        }
+    CompletableFuture<Response> future = new CompletableFuture<>();
+    persistence()
+        .executeAuthResponse(
+            () -> {
+              try {
+                Authenticator.SaslNegotiator negotiator =
+                    ((ServerConnection) connection).getSaslNegotiator();
+                byte[] challenge = negotiator.evaluateResponse(token);
+                if (negotiator.isComplete()) {
+                  AuthenticatedUser authenticatedUser = negotiator.getAuthenticatedUser();
+                  persistenceConnection().login(authenticatedUser);
+                  if (authenticatedUser.token() != null) {
+                    ((ServerConnection) connection)
+                        .clientInfo()
+                        .setAuthToken(authenticatedUser.token());
+                  }
+
+                  ClientMetrics.instance.markAuthSuccess();
+                  // authentication is complete, send a ready message to the client
 
-        ClientMetrics.instance.markAuthSuccess();
-        // authentication is complete, send a ready message to the client
-        return CompletableFuture.completedFuture(new AuthSuccess(challenge));
-      } else {
-        return CompletableFuture.completedFuture(new AuthChallenge(challenge));
-      }
-    } catch (AuthenticationException e) {
-      ClientMetrics.instance.markAuthFailure();
-      return CompletableFuture.completedFuture(ErrorMessage.fromException(e));
-    }
+                  future.complete(new AuthSuccess(challenge));
+                } else {
+                  future.complete(new AuthChallenge(challenge));
+                }
+              } catch (AuthenticationException e) {
```

Comment:
should the returned `CompletableFuture` wrap all possible exceptions? If there will be a different kind of exception, it will be propagated to the caller now. I think that when returning the `CompletableFuture` it should encapsulate all kinds of exceptions, so I suggest to change this code to:
```
  } catch (Exception e) {
  // ...
  }
```

Reply:
Good point. I missed that :facepalm:

Reply:
This is already handle at the `.execute()` call site: https://github.com/stargate/stargate/blob/v1.0.0/cql/src/main/java/org/apache/cassandra/stargate/transport/internal/Message.java#L715-L717

The `} catch (AuthenticationException e) {` is to specifically handle and mark failed authentication attempts.

Reply:
I'm not sure... if an unexpected exception occurs within the task submitted to `persistence().executeAuthResponse(...)` the future retuned from this method will never complete, will it?

Reply:
yeah, you're right.

Reply:
Added fix: https://github.com/stargate/stargate/pull/530/commits/bab6c86b912184c046ff6580dd306207a02a9dae

Reply:
Good catch. Thanks.